File: db/crud/catalog.py
"""
Catalog and Search CRUD operations for Stremio API.

These functions handle catalog queries, MDBList integration, and search
for the Stremio addon endpoints.

ARCHITECTURE NOTE:
- Internally, all queries use media_id (integer) for performance
- External ID translation (to IMDb/TMDB) happens only at the Stremio boundary
- get_canonical_external_ids_batch() is used for batch translation
"""

import logging
from typing import Literal

from fastapi import BackgroundTasks
from sqlalchemy import asc, desc, func, union
from sqlmodel import select
from sqlmodel.ext.asyncio.session import AsyncSession

from db import public_schemas, schemas
from db.crud.media import get_canonical_external_ids_batch
from db.enums import MediaType
from db.models import (
    AkaTitle,
    Catalog,
    Genre,
    Media,
    MediaCatalogLink,
    MediaExternalID,
    MediaGenreLink,
    MediaParentalCertificateLink,
    MediaRating,
    MovieMetadata,
    ParentalCertificate,
    RatingProvider,
    SeriesMetadata,
    Stream,
    StreamMediaLink,
    TorrentStream,
    TVMetadata,
    UserLibraryItem,
)
from db.schemas import UserData

logger = logging.getLogger(__name__)

# Type alias for catalog sort options
CatalogSortOption = Literal["latest", "popular", "rating", "year", "title", "release_date"]
MY_LIBRARY_CATALOG_TYPE_MAP: dict[str, MediaType] = {
    "my_library_movies": MediaType.MOVIE,
    "my_library_series": MediaType.SERIES,
    "my_library_tv": MediaType.TV,
}


async def get_catalog_meta_list(
    session: AsyncSession,
    catalog_type: MediaType,
    catalog_id: str,
    user_data: UserData,
    skip: int = 0,
    limit: int = 25,
    genre: str | None = None,
    namespace: str | None = None,
    is_watchlist_catalog: bool = False,
    info_hashes: list[str] | None = None,
    sort: CatalogSortOption | None = None,
    sort_dir: Literal["asc", "desc"] | None = None,
) -> public_schemas.Metas:
    """
    Get metadata list for catalog with efficient filtering and pagination.

    Args:
        session: Database session
        catalog_type: Type of media (movie, series, tv)
        catalog_id: Catalog identifier
        user_data: User preferences and filters
        skip: Pagination offset
        limit: Max results to return
        genre: Optional genre filter
        namespace: Optional namespace for TV (deprecated in v5)
        is_watchlist_catalog: Whether this is a watchlist query
        info_hashes: List of torrent info hashes for watchlist queries
        sort: Sort field. Options: latest, popular, rating, year, title, release_date.
              Defaults to "latest" if not specified.
        sort_dir: Sort direction. Options: asc, desc. Defaults to "desc".

    Returns:
        Metas object containing list of Meta items for Stremio
    """
    # Build base query - use media.id internally, external_id translation happens at Stremio boundary
    query = select(Media.id, Media.title).where(Media.type == catalog_type)
    is_my_library_catalog = catalog_id in MY_LIBRARY_CATALOG_TYPE_MAP

    # Handle personal My Library catalogs
    if is_my_library_catalog:
        if not user_data.user_id:
            return public_schemas.Metas(metas=[])

        expected_type = MY_LIBRARY_CATALOG_TYPE_MAP[catalog_id]
        if catalog_type != expected_type:
            return public_schemas.Metas(metas=[])

        query = query.join(UserLibraryItem, UserLibraryItem.media_id == Media.id).where(
            UserLibraryItem.user_id == user_data.user_id
        )
    # Handle watchlist catalog
    elif is_watchlist_catalog and info_hashes:
        # Join through TorrentStream to find media by info_hash
        query = (
            query.join(StreamMediaLink, StreamMediaLink.media_id == Media.id)
            .join(Stream, Stream.id == StreamMediaLink.stream_id)
            .join(TorrentStream, TorrentStream.stream_id == Stream.id)
            .where(TorrentStream.info_hash.in_([h.lower() for h in info_hashes]))
        )
    else:
        # Standard catalog filter
        if catalog_type != MediaType.TV:
            query = (
                query.join(MediaCatalogLink, MediaCatalogLink.media_id == Media.id)
                .join(Catalog, Catalog.id == MediaCatalogLink.catalog_id)
                .where(Catalog.name == catalog_id)
            )

    # Add content filters for movies and series
    if catalog_type in [MediaType.MOVIE, MediaType.SERIES] and not is_my_library_catalog:
        specific_model = MovieMetadata if catalog_type == MediaType.MOVIE else SeriesMetadata
        query = query.join(specific_model, specific_model.media_id == Media.id)

        # Apply nudity filter
        if "Disable" not in user_data.nudity_filter:
            query = query.where(Media.nudity_status.notin_(user_data.nudity_filter))

        # Apply certification filter
        if "Disable" not in user_data.certification_filter:
            blocked_cert_exists = (
                select(MediaParentalCertificateLink.media_id)
                .join(ParentalCertificate)
                .where(
                    MediaParentalCertificateLink.media_id == Media.id,
                    ParentalCertificate.name.in_(user_data.certification_filter),
                )
                .exists()
            )
            query = query.where(~blocked_cert_exists)

    # Add TV-specific filters
    if catalog_type == MediaType.TV and not is_my_library_catalog:
        query = (
            query.join(TVMetadata, TVMetadata.media_id == Media.id)
            .join(StreamMediaLink, StreamMediaLink.media_id == Media.id)
            .join(Stream, Stream.id == StreamMediaLink.stream_id)
            .where(Stream.is_active.is_(True), Stream.is_blocked.is_(False))
        )

    # Add genre filter
    if genre:
        query = (
            query.join(MediaGenreLink, MediaGenreLink.media_id == Media.id)
            .join(Genre, Genre.id == MediaGenreLink.genre_id)
            .where(Genre.name == genre)
        )

    # Apply sorting based on user preferences
    sort = sort or "latest"
    sort_dir = sort_dir or "desc"
    order_func = asc if sort_dir == "asc" else desc
    nulls_position = "nulls_first" if sort_dir == "asc" else "nulls_last"

    if sort == "latest":
        order_expr = (
            order_func(UserLibraryItem.added_at) if is_my_library_catalog else order_func(Media.last_stream_added)
        )
        query = query.order_by(getattr(order_expr, nulls_position)())
    elif sort in ("popular", "rating"):
        # Sort by IMDb rating with fallback to total_streams
        imdb_rating_subq = (
            select(MediaRating.rating)
            .join(RatingProvider, RatingProvider.id == MediaRating.rating_provider_id)
            .where(
                MediaRating.media_id == Media.id,
                RatingProvider.name == "imdb",
            )
            .correlate(Media)
            .scalar_subquery()
        )
        rating_order = order_func(imdb_rating_subq)
        streams_order = order_func(Media.total_streams)
        query = query.order_by(
            getattr(rating_order, nulls_position)(),
            getattr(streams_order, nulls_position)(),
        )
    elif sort == "year":
        order_expr = order_func(Media.year)
        query = query.order_by(getattr(order_expr, nulls_position)())
    elif sort == "release_date":
        order_expr = order_func(Media.release_date)
        query = query.order_by(getattr(order_expr, nulls_position)())
    elif sort == "title":
        query = query.order_by(order_func(Media.title))
    else:
        # Fallback to latest
        order_expr = order_func(Media.last_stream_added)
        query = query.order_by(getattr(order_expr, nulls_position)())

    # Apply pagination
    query = query.offset(skip).limit(limit)

    result = await session.exec(query)
    data = result.unique().all()

    if not data:
        return public_schemas.Metas(metas=[])

    # Extract media_ids and titles - data is (media_id, title) tuples
    media_ids = [row[0] for row in data]
    titles_map = {row[0]: row[1] for row in data}

    # Batch translate media_ids to external_ids for Stremio response
    external_ids = await get_canonical_external_ids_batch(session, media_ids)

    # Build Stremio response with external_ids
    metas = [
        public_schemas.Meta(
            id=external_ids.get(media_id, f"mf:{media_id}"),
            name=titles_map[media_id],
            type=catalog_type,
        )
        for media_id in media_ids
    ]
    return public_schemas.Metas(metas=metas)


async def get_mdblist_meta_list(
    session: AsyncSession,
    user_data: UserData,
    background_tasks: BackgroundTasks,
    list_config: schemas.MDBListItem,
    catalog_type: str,
    genre: str | None = None,
    skip: int = 0,
    limit: int = 50,
) -> list[public_schemas.Meta]:
    """
    Get a list of metadata entries from MDBList.

    This function:
    1. Initializes the MDBList scraper with user's API key
    2. If not using filters, returns items directly from MDBList
    3. If using filters, fetches IMDb IDs and filters through PostgreSQL
       with user's parental guide filters applied
    4. Triggers background fetch for missing metadata

    Args:
        session: Database session
        user_data: User data with MDBList config
        background_tasks: FastAPI background tasks for async operations
        list_config: MDBList configuration for the specific list
        catalog_type: 'movie' or 'series'
        genre: Optional genre filter
        skip: Pagination offset
        limit: Max results to return

    Returns:
        List of Meta objects for Stremio
    """

    # Lazy import to avoid circular dependencies
    from scrapers.mdblist import initialize_mdblist_scraper

    if not user_data.mdblist_config:
        return []

    # Determine media type enum
    media_type_enum = MediaType.MOVIE if catalog_type == "movie" else MediaType.SERIES

    # Initialize MDBList scraper
    mdblist_scraper = await initialize_mdblist_scraper(user_data.mdblist_config.api_key)
    try:
        if not list_config.use_filters:
            # Return items directly from MDBList without filtering
            mdblist_items = await mdblist_scraper.get_list_items(
                list_config=list_config,
                skip=skip,
                limit=limit,
                genre=genre,
                use_filters=False,
            )
            # Convert to public_schemas.Meta for compatibility
            return [
                public_schemas.Meta(
                    id=item.id,
                    name=item.name,
                    type=item.type,
                    poster=item.poster,
                )
                for item in mdblist_items
            ]

        # For filtered results, get all IMDb IDs first
        imdb_ids = await mdblist_scraper.get_list_items(
            list_config=list_config,
            skip=0,
            limit=0,  # Ignored for filtered results
            genre=genre,
            use_filters=True,
        )

        if not imdb_ids:
            return []

        # Build PostgreSQL query with filters
        specific_model = MovieMetadata if media_type_enum == MediaType.MOVIE else SeriesMetadata

        # Build base query - join with MediaExternalID to match IMDb IDs
        # We select both media_id (for internal use) and external_id (for Stremio response)
        query = (
            select(Media.id, Media.title, MediaExternalID.external_id)
            .join(MediaExternalID, MediaExternalID.media_id == Media.id)
            .where(
                MediaExternalID.provider == "imdb",
                MediaExternalID.external_id.in_(imdb_ids),
                Media.type == media_type_enum,
                Media.total_streams > 0,
            )
            .join(specific_model, specific_model.media_id == Media.id)
        )

        # Apply nudity filter
        if "Disable" not in user_data.nudity_filter:
            query = query.where(Media.nudity_status.notin_(user_data.nudity_filter))

        # Apply certification filter
        if "Disable" not in user_data.certification_filter:
            blocked_cert_exists = (
                select(MediaParentalCertificateLink.media_id)
                .join(ParentalCertificate)
                .where(
                    MediaParentalCertificateLink.media_id == Media.id,
                    ParentalCertificate.name.in_(user_data.certification_filter),
                )
                .exists()
            )
            query = query.where(~blocked_cert_exists)

        # Add sorting and pagination
        query = query.order_by(Media.last_stream_added.desc().nullslast()).offset(skip).limit(limit)

        result = await session.exec(query)
        data = result.unique().all()

        if not data:
            # Check for missing metadata and trigger background fetch
            existing_query = select(MediaExternalID.external_id).where(
                MediaExternalID.provider == "imdb",
                MediaExternalID.external_id.in_(imdb_ids),
            )
            existing_result = await session.exec(existing_query)
            existing_ids = set(existing_result.all())
            missing_ids = list(set(imdb_ids) - existing_ids)

            # TODO: Implement batch metadata fetching for missing IDs
            # if missing_ids:
            #     background_tasks.add_task(fetch_metadata_batch, missing_ids, catalog_type)
            if missing_ids:
                logger.debug(f"Found {len(missing_ids)} missing metadata IDs")
            return []

        # Convert to Meta objects - data is (media_id, title, external_id)
        metas = [
            public_schemas.Meta(id=row[2], name=row[1], type=media_type_enum)
            for row in data
            if row[2]  # Skip entries without external_id
        ]
        return metas

    finally:
        await mdblist_scraper.close()


async def search_metadata(
    session: AsyncSession,
    catalog_type: MediaType,
    search_query: str,
    user_data: UserData,
    namespace: str | None = None,
    limit: int = 50,
) -> public_schemas.Metas:
    """
    Search metadata with efficient filtering and ranking.

    Uses PostgreSQL full-text search (GIN index on title_tsv) and
    trigram search (gin_trgm_ops) for fuzzy matching.

    Args:
        session: Database session
        catalog_type: Type of media to search
        search_query: Search text
        user_data: User preferences for filtering
        namespace: Optional namespace for TV (deprecated in v5)
        limit: Max results to return

    Returns:
        Metas object containing matching Meta items
    """
    search_query = search_query.strip()
    if not search_query:
        return public_schemas.Metas(metas=[])

    # Create search vector for full-text search
    search_vector = func.plainto_tsquery("simple", search_query.lower())

    # Phase 1: Full-text search matches (fastest - uses GIN index on title_tsv)
    fts_base_matches = (
        select(Media.id)
        .where(
            Media.title_tsv.op("@@")(search_vector),
            Media.type == catalog_type,
        )
        .limit(200)
    )

    fts_aka_matches = select(AkaTitle.media_id).where(AkaTitle.title_tsv.op("@@")(search_vector)).limit(200)

    # Phase 2: Trigram search using % operator (uses GIN index with gin_trgm_ops)
    trgm_base_matches = (
        select(Media.id)
        .where(
            Media.title.op("%")(search_query),
            Media.type == catalog_type,
        )
        .limit(100)
    )

    # Combine matches using UNION (deduplicates results)
    all_matches = union(fts_base_matches, fts_aka_matches, trgm_base_matches).subquery()

    # Build main query - use media_id internally, external_id translation at the end
    query = select(Media.id, Media.title).where(Media.type == catalog_type, Media.id.in_(select(all_matches.c.id)))

    # Add content filters for movies and series
    if catalog_type in [MediaType.MOVIE, MediaType.SERIES]:
        specific_model = MovieMetadata if catalog_type == MediaType.MOVIE else SeriesMetadata
        query = query.join(specific_model, specific_model.media_id == Media.id)

        # Apply nudity filter
        if "Disable" not in user_data.nudity_filter:
            query = query.where(Media.nudity_status.notin_(user_data.nudity_filter))

        # Apply certification filter
        if "Disable" not in user_data.certification_filter:
            blocked_cert_exists = (
                select(MediaParentalCertificateLink.media_id)
                .join(ParentalCertificate)
                .where(
                    MediaParentalCertificateLink.media_id == Media.id,
                    ParentalCertificate.name.in_(user_data.certification_filter),
                )
                .exists()
            )
            query = query.where(~blocked_cert_exists)

        # Add stream existence filter for movies/series
        stream_exists = (
            select(StreamMediaLink.id)
            .join(Stream, Stream.id == StreamMediaLink.stream_id)
            .where(
                StreamMediaLink.media_id == Media.id,
                Stream.is_active.is_(True),
                Stream.is_blocked.is_(False),
            )
            .exists()
        )
        query = query.where(stream_exists)

    # Add TV-specific filters
    if catalog_type == MediaType.TV:
        query = (
            query.join(TVMetadata, TVMetadata.media_id == Media.id)
            .join(StreamMediaLink, StreamMediaLink.media_id == Media.id)
            .join(Stream, Stream.id == StreamMediaLink.stream_id)
            .where(Stream.is_active.is_(True), Stream.is_blocked.is_(False))
        )

    # Add ordering by relevance and limit
    query = query.order_by(func.ts_rank_cd(Media.title_tsv, search_vector).desc(), Media.title).limit(limit)

    result = await session.exec(query)
    data = result.unique().all()

    if not data:
        return public_schemas.Metas(metas=[])

    # Extract media_ids and titles - data is (media_id, title) tuples
    media_ids = [row[0] for row in data]
    titles_map = {row[0]: row[1] for row in data}

    # Batch translate media_ids to external_ids for Stremio response
    external_ids = await get_canonical_external_ids_batch(session, media_ids)

    # Build Stremio response with external_ids
    metas = [
        public_schemas.Meta(
            id=external_ids.get(media_id, f"mf:{media_id}"),
            name=titles_map[media_id],
            type=catalog_type,
        )
        for media_id in media_ids
    ]
    return public_schemas.Metas(metas=metas)

File: backend/src/db/catalog.rs
use sqlx::PgPool;
use tracing::warn;

use super::types::{nudity_statuses_from_filter, MediaId, MediaType, UserId};

const LIMIT: i64 = 100;
const WATCHLIST_LIMIT: i64 = 25;

#[derive(sqlx::FromRow, Debug)]
pub struct CatalogRow {
    pub media_id: MediaId,
    pub media_type: MediaType,
    pub title: String,
    pub year: Option<i32>,
    pub end_year: Option<i32>,
    pub description: Option<String>,
    pub imdb_id: Option<String>,
    pub poster_url: Option<String>,
}

fn order_clause(sort: &str, sort_dir: &str) -> String {
    let sort_dir = if sort_dir == "asc" { "ASC" } else { "DESC" };
    let nulls = if sort_dir == "ASC" {
        "NULLS FIRST"
    } else {
        "NULLS LAST"
    };
    match sort {
        "popular" => format!(
            "m.popularity {sort_dir} {nulls}, m.total_streams {sort_dir} {nulls}, m.last_stream_added {sort_dir} {nulls}, m.id ASC"
        ),
        "rating" => format!(
            "(SELECT mr2.rating FROM media_rating mr2 JOIN rating_provider rp2 ON rp2.id = mr2.rating_provider_id WHERE mr2.media_id = m.id AND rp2.name = 'imdb' LIMIT 1) {sort_dir} {nulls}, m.total_streams {sort_dir} {nulls}, m.id ASC"
        ),
        "year" => format!("m.year {sort_dir} {nulls}, m.id ASC"),
        "title" => format!("m.title {sort_dir}, m.id ASC"),
        "release_date" => format!(
            "COALESCE(m.release_date, make_date(m.year, 12, 31)) {sort_dir} {nulls}, m.id ASC"
        ),
        _ => format!("m.last_stream_added {sort_dir} {nulls}, m.id ASC"),
    }
}

pub struct CatalogQuery<'a> {
    pub catalog_id: &'a str,
    pub media_type: &'a str,
    pub skip: i64,
    pub genre: Option<&'a str>,
    pub nudity_excludes: &'a [String],
    /// Parental certificate names to exclude (e.g. ["Adults+"]). Empty = no filter.
    pub cert_excludes: &'a [String],
    pub sort: &'a str,
    pub sort_dir: &'a str,
    /// Set for my_library_* catalogs; filters by user_library_item.user_id.
    pub user_id: Option<UserId>,
}

pub async fn get_catalog_items(pool: &PgPool, q: CatalogQuery<'_>) -> Vec<CatalogRow> {
    let CatalogQuery {
        catalog_id,
        media_type,
        skip,
        genre,
        nudity_excludes,
        cert_excludes,
        sort,
        sort_dir,
        user_id,
    } = q;

    let Some(mt) = MediaType::from_wire(media_type) else {
        return vec![];
    };

    // my_library_* catalogs join through user_library_item instead of catalog/media_catalog_link.
    if catalog_id.starts_with("my_library_") {
        return get_library_items(pool, mt, skip, nudity_excludes, cert_excludes, user_id).await;
    }

    let nudity_exclude_enums = nudity_statuses_from_filter(nudity_excludes);
    let ord = order_clause(sort, sort_dir);
    let sql = format!(
        r#"
        SELECT
            m.id AS media_id,
            m.type AS media_type,
            m.title,
            m.year,
            EXTRACT(YEAR FROM m.end_date)::int AS end_year,
            m.description,
            mei.external_id AS imdb_id,
            mi.url AS poster_url
        FROM media m
        JOIN media_catalog_link mcl ON mcl.media_id = m.id
        JOIN catalog c ON c.id = mcl.catalog_id AND c.name = $1
        LEFT JOIN media_external_id mei ON mei.media_id = m.id AND mei.provider = 'imdb'
        LEFT JOIN LATERAL (
            SELECT url FROM media_image
            WHERE media_id = m.id AND image_type = 'poster' AND is_primary = true
            LIMIT 1
        ) mi ON true
        WHERE m.type = $2
          AND m.total_streams > 0
          AND NOT m.is_blocked
          AND ($4::text IS NULL OR EXISTS (
              SELECT 1 FROM media_genre_link mgl
              JOIN genre g ON g.id = mgl.genre_id
              WHERE mgl.media_id = m.id AND g.name = $4
          ))
          AND (cardinality($5::nuditystatus[]) = 0 OR m.nudity_status <> ALL($5))
          AND (cardinality($6::text[]) IS NULL OR NOT EXISTS (
              SELECT 1 FROM media_parental_certificate_link mpcl
              JOIN parental_certificate pc ON pc.id = mpcl.certificate_id
              WHERE mpcl.media_id = m.id AND pc.name = ANY($6)
          ))
        ORDER BY {ord}
        LIMIT {LIMIT} OFFSET $3
        "#
    );

    sqlx::query_as::<_, CatalogRow>(&sql)
        .bind(catalog_id) // $1
        .bind(mt) // $2
        .bind(skip) // $3
        .bind(genre) // $4 - Option<&str> → NULL when None
        .bind(&nudity_exclude_enums) // $5 - nuditystatus[]
        .bind(cert_excludes) // $6 - &[String] → text[]
        .fetch_all(pool)
        .await
        .unwrap_or_else(|e| {
            warn!("catalog query [{catalog_id}]: {e}");
            vec![]
        })
}

async fn get_library_items(
    pool: &PgPool,
    media_type: MediaType,
    skip: i64,
    nudity_excludes: &[String],
    cert_excludes: &[String],
    user_id: Option<UserId>,
) -> Vec<CatalogRow> {
    let Some(uid) = user_id else {
        return vec![];
    };
    let nudity_exclude_enums = nudity_statuses_from_filter(nudity_excludes);
    sqlx::query_as::<_, CatalogRow>(
        r#"
        SELECT
            m.id AS media_id,
            m.type AS media_type,
            m.title,
            m.year,
            EXTRACT(YEAR FROM m.end_date)::int AS end_year,
            m.description,
            mei.external_id AS imdb_id,
            mi.url AS poster_url
        FROM user_library_item uli
        JOIN media m ON m.id = uli.media_id
        LEFT JOIN media_external_id mei ON mei.media_id = m.id AND mei.provider = 'imdb'
        LEFT JOIN LATERAL (
            SELECT url FROM media_image
            WHERE media_id = m.id AND image_type = 'poster' AND is_primary = true
            LIMIT 1
        ) mi ON true
        WHERE uli.user_id = $1
          AND m.type = $2
          AND NOT m.is_blocked
          AND (cardinality($3::nuditystatus[]) = 0 OR m.nudity_status <> ALL($3))
          AND (cardinality($5::text[]) IS NULL OR NOT EXISTS (
              SELECT 1 FROM media_parental_certificate_link mpcl
              JOIN parental_certificate pc ON pc.id = mpcl.certificate_id
              WHERE mpcl.media_id = m.id AND pc.name = ANY($5)
          ))
        ORDER BY uli.added_at DESC
        LIMIT 100 OFFSET $4
        "#,
    )
    .bind(uid) // $1
    .bind(media_type) // $2
    .bind(&nudity_exclude_enums) // $3
    .bind(skip) // $4
    .bind(cert_excludes) // $5
    .fetch_all(pool)
    .await
    .unwrap_or_else(|e| {
        warn!(
            "library query [uid={uid} type={}]: {e}",
            media_type.as_wire()
        );
        vec![]
    })
}

/// Watchlist catalog rows: media linked to any of the given torrent info_hashes.
pub async fn get_watchlist_items(
    pool: &PgPool,
    media_type: &str,
    info_hashes: &[String],
    skip: i64,
    nudity_excludes: &[String],
    cert_excludes: &[String],
    sort: &str,
    sort_dir: &str,
) -> Vec<CatalogRow> {
    if info_hashes.is_empty() {
        return vec![];
    }

    let Some(mt) = MediaType::from_wire(media_type) else {
        return vec![];
    };

    let info_hashes_lower: Vec<String> = info_hashes.iter().map(|h| h.to_lowercase()).collect();
    let nudity_exclude_enums = nudity_statuses_from_filter(nudity_excludes);
    let ord = order_clause(sort, sort_dir);
    let sql = format!(
        r#"
        SELECT
            m.id AS media_id,
            m.type AS media_type,
            m.title,
            m.year,
            EXTRACT(YEAR FROM m.end_date)::int AS end_year,
            m.description,
            mei.external_id AS imdb_id,
            mi.url AS poster_url
        FROM media m
        JOIN stream_media_link sml ON sml.media_id = m.id
        JOIN stream s ON s.id = sml.stream_id
        JOIN torrent_stream ts ON ts.stream_id = s.id
        LEFT JOIN media_external_id mei ON mei.media_id = m.id AND mei.provider = 'imdb'
        LEFT JOIN LATERAL (
            SELECT url FROM media_image
            WHERE media_id = m.id AND image_type = 'poster' AND is_primary = true
            LIMIT 1
        ) mi ON true
        WHERE m.type = $1
          AND m.total_streams > 0
          AND NOT m.is_blocked
          AND lower(ts.info_hash) = ANY($2)
          AND (cardinality($3::nuditystatus[]) = 0 OR m.nudity_status <> ALL($3))
          AND (cardinality($4::text[]) IS NULL OR NOT EXISTS (
              SELECT 1 FROM media_parental_certificate_link mpcl
              JOIN parental_certificate pc ON pc.id = mpcl.certificate_id
              WHERE mpcl.media_id = m.id AND pc.name = ANY($4)
          ))
        GROUP BY m.id, m.type, m.title, m.year, m.end_date, m.description, mei.external_id, mi.url
        ORDER BY {ord}
        LIMIT {WATCHLIST_LIMIT} OFFSET $5
        "#
    );

    sqlx::query_as::<_, CatalogRow>(&sql)
        .bind(mt) // $1
        .bind(&info_hashes_lower) // $2
        .bind(&nudity_exclude_enums) // $3
        .bind(cert_excludes) // $4
        .bind(skip) // $5
        .fetch_all(pool)
        .await
        .unwrap_or_else(|e| {
            warn!("watchlist catalog query [type={media_type}]: {e}");
            vec![]
        })
}

pub async fn search_metadata(
    pool: &PgPool,
    media_type: &str,
    query: &str,
    skip: i64,
    nudity_excludes: &[String],
    cert_excludes: &[String],
) -> Vec<CatalogRow> {
    let Some(mt) = MediaType::from_wire(media_type) else {
        return vec![];
    };

    let nudity_exclude_enums = nudity_statuses_from_filter(nudity_excludes);
    sqlx::query_as::<_, CatalogRow>(
        r#"
        SELECT
            m.id AS media_id,
            m.type AS media_type,
            m.title,
            m.year,
            EXTRACT(YEAR FROM m.end_date)::int AS end_year,
            m.description,
            mei.external_id AS imdb_id,
            mi.url AS poster_url
        FROM media m
        LEFT JOIN media_external_id mei ON mei.media_id = m.id AND mei.provider = 'imdb'
        LEFT JOIN LATERAL (
            SELECT url FROM media_image
            WHERE media_id = m.id AND image_type = 'poster' AND is_primary = true
            LIMIT 1
        ) mi ON true
        WHERE m.type = $1
          AND m.total_streams > 0
          AND NOT m.is_blocked
          AND (cardinality($3::nuditystatus[]) = 0 OR m.nudity_status <> ALL($3))
          AND (cardinality($5::text[]) IS NULL OR NOT EXISTS (
              SELECT 1 FROM media_parental_certificate_link mpcl
              JOIN parental_certificate pc ON pc.id = mpcl.certificate_id
              WHERE mpcl.media_id = m.id AND pc.name = ANY($5)
          ))
          AND m.id IN (
              -- Phase 1: FTS on main title (uses GIN index)
              SELECT m2.id FROM media m2
              WHERE m2.title_tsv @@ plainto_tsquery('simple', $2)
                AND m2.type = $1
              UNION
              -- Phase 2: FTS on aka_title
              SELECT at2.media_id FROM aka_title at2
              WHERE at2.title_tsv @@ plainto_tsquery('simple', $2)
              UNION
              -- Phase 3: trigram similarity (uses GIN gin_trgm_ops index)
              SELECT m3.id FROM media m3
              WHERE m3.title % $2
                AND m3.type = $1
          )
        ORDER BY
            ts_rank_cd(m.title_tsv, plainto_tsquery('simple', $2)) DESC,
            m.total_streams DESC
        LIMIT 100 OFFSET $4
        "#,
    )
    .bind(mt) // $1
    .bind(query) // $2
    .bind(&nudity_exclude_enums) // $3
    .bind(skip) // $4
    .bind(cert_excludes) // $5
    .fetch_all(pool)
    .await
    .unwrap_or_else(|e| {
        warn!("search query [type={media_type} q={query}]: {e}");
        vec![]
    })
}

/// Filter MDBList IMDb ids through PostgreSQL with parental-guide filters and stream availability.
pub async fn get_mdblist_filtered_items(
    pool: &PgPool,
    media_type: MediaType,
    imdb_ids: &[String],
    skip: i64,
    limit: i64,
    nudity_excludes: &[String],
    cert_excludes: &[String],
) -> Vec<CatalogRow> {
    if imdb_ids.is_empty() {
        return vec![];
    }

    let nudity_exclude_enums = nudity_statuses_from_filter(nudity_excludes);
    let type_join = match media_type {
        MediaType::Movie => "JOIN movie_metadata mm ON mm.media_id = m.id",
        MediaType::Series => "JOIN series_metadata sm ON sm.media_id = m.id",
        _ => return vec![],
    };

    let sql = format!(
        r#"
        SELECT
            m.id AS media_id,
            m.type AS media_type,
            m.title,
            m.year,
            EXTRACT(YEAR FROM m.end_date)::int AS end_year,
            m.description,
            mei.external_id AS imdb_id,
            mi.url AS poster_url
        FROM media m
        JOIN media_external_id mei ON mei.media_id = m.id AND mei.provider = 'imdb'
        {type_join}
        LEFT JOIN LATERAL (
            SELECT url FROM media_image
            WHERE media_id = m.id AND image_type = 'poster' AND is_primary = true
            LIMIT 1
        ) mi ON true
        WHERE mei.external_id = ANY($1)
          AND m.type = $2
          AND m.total_streams > 0
          AND (cardinality($3::nuditystatus[]) = 0 OR m.nudity_status <> ALL($3))
          AND (cardinality($4::text[]) IS NULL OR NOT EXISTS (
              SELECT 1 FROM media_parental_certificate_link mpcl
              JOIN parental_certificate pc ON pc.id = mpcl.certificate_id
              WHERE mpcl.media_id = m.id AND pc.name = ANY($4)
          ))
        ORDER BY m.last_stream_added DESC NULLS LAST, m.id ASC
        LIMIT $5 OFFSET $6
        "#
    );

    sqlx::query_as::<_, CatalogRow>(&sql)
        .bind(imdb_ids)
        .bind(media_type)
        .bind(&nudity_exclude_enums)
        .bind(cert_excludes)
        .bind(limit)
        .bind(skip)
        .fetch_all(pool)
        .await
        .unwrap_or_else(|e| {
            warn!(
                "mdblist filtered query [type={}]: {e}",
                media_type.as_wire()
            );
            vec![]
        })
}
